src/Workspaces/Core/Portable/Formatting/Engine/TokenStream.cs

```diff
@@ -54,19 +56,16 @@ public TokenStream(TreeData treeData, OptionSet optionSet, TextSpan spanToFormat
                 _treeData = treeData;
                 _optionSet = optionSet;
 
-                // use some heuristics to get initial size of list rather than blindly start from default size == 4
-                int sizeOfList = spanToFormat.Length / MagicTextLengthToTokensRatio;
-                _tokens = new List<SyntaxToken>(sizeOfList);
-                _tokens.AddRange(_treeData.GetApplicableTokens(spanToFormat));
+                _tokens = _treeData.GetApplicableTokens(spanToFormat).ToImmutableArrayOrEmpty();
+                // estimate average token length to speed up searches for token index in the stream
+                int approximateStreamLength = _tokens[_tokens.Length - 1].FullSpan.Start - _tokens[0].FullSpan.Start;
```

Comment:
1. you can say `tokens.Last()`.  It's O(1) and non-allocating for immutable-arrays.
2. are we certain this array is non-empty?
3. in general we use 'var', but i can see this file is not consistent. so it's up to you which you'd like to follow here.

Reply:
Shouldn't this `_tokens.Last().FullSpan.End - tokens[0].FullSpan.Start`?  I would use the *end* of the last token, not the start of it.